### PR TITLE
Improve etcd-client cache stability

### PR DESCRIPTION
**Pull request by pzheng1025**
- clean cache when error happens
- Add shutdown for kv client and use lazy init

### CODE REVIEW

File: etcd-client/src/kv/mod.rs [hunk 1]
```diff
@@ -58,18 +65,21 @@ pub struct Kv {
     cache_enable: bool,
     /// Etcd watch request sender.
     watch_sender: Sender<EtcdWatchRequest>,
-    /// Flag to indicate if the cache is error and disabled
-    cache_err: AtomicBool,
+    /// Read write lock to indicate the availability of the cache.
+    /// When the cache has error, write lock is acquired to disable the cache and clean the cache.
+    /// Other operations will get read lock.
+    cache_availability: RwLock<()>,
+    /// A channel sender to send shutdowm request.
```
Comment (mrwwheat7) [inline etcd-client/src/kv/mod.rs]: shutdown

File: etcd-client/src/kv/mod.rs [hunk 2]
```diff
@@ -175,18 +203,55 @@ impl Kv {
...
+    ///
+    /// This method should only be called within etcd client.
+    pub(crate) fn new(
+        client: KvClient,
+        watch_client: WatchClient,
+        cache_size: usize,
+        cache_enable: bool,
+    ) -> Self {
+        Self {
+            inner: Arc::new(Lazy::new(move || {
```
Comment (mrwwheat7) [inline etcd-client/src/kv/mod.rs:250]: Why do we need Lazy here?

File: etcd-client/src/kv/mod.rs [hunk 3]
```diff
@@ -195,26 +260,32 @@ impl Kv {
...
+                    let revision = resp.get_revision();
+                    // Creates a new watch request and adds to the send queue.
+                    let mut watch_request = EtcdWatchRequest::create(KeyRange::key(key));
+                    watch_request.set_start_revision(revision.cast());
+                    if let Err(e) = inner.watch_sender.send(watch_request).await {
+                        warn!(
+                            "Fail to send watch request, the error is {}, clean cache",
+                            e
+                        );
+                        drop(cache_lock);
```
Comment (mrwwheat7) [inline etcd-client/src/kv/mod.rs:282]: Do we need to drop the lock manually here?

File: etcd-client/src/kv/mod.rs (hunk 1 above)
Comment (pzheng1025) [reply etcd-client/src/kv/mod.rs]: Done

File: etcd-client/src/kv/mod.rs (hunk 2 above)
Comment (pzheng1025) [reply etcd-client/src/kv/mod.rs:250]: If users don't use kv-client like watch, lock etc, don't create smol task. Lazy will create kv client when users use it first time.

File: etcd-client/src/kv/mod.rs (hunk 3 above)
Comment (pzheng1025) [reply etcd-client/src/kv/mod.rs:282]: Because clean_cache will get write lock, so drop read lock here.

File: etcd-client/src/client.rs [hunk 4]
```diff
@@ -142,7 +142,7 @@ impl Client {
                 auth_client: Auth::new(AuthClient::new(channel.clone())),
                 kv_client: Kv::new(
                     KvClient::new(channel.clone()),
-                    &etcd_watch_client,
+                    etcd_watch_client.clone(),
```
Comment (pwang7) [inline etcd-client/src/client.rs:145]: Why clone here?

Comment (pzheng1025) [reply etcd-client/src/client.rs:145]: Because Lazy::new() init requires static lifetime, so need to clone and give ownership.

**Review by mrwwheat7 — APPROVED** [review]

**Review by mrwwheat7 — COMMENTED** [review]
So looks like missing `shut_down` function is the reason for the failure in etcd client, right?

File: etcd-client/src/kv/mod.rs [hunk 5]
```diff
@@ -58,18 +65,21 @@ pub struct Kv {
     cache_enable: bool,
     /// Etcd watch request sender.
     watch_sender: Sender<EtcdWatchRequest>,
-    /// Flag to indicate if the cache is error and disabled
-    cache_err: AtomicBool,
+    /// Read write lock to indicate the availability of the cache.
+    /// When the cache has error, write lock is acquired to disable the cache and clean the cache.
+    /// Other operations will get read lock.
+    cache_availability: RwLock<()>,
```
Comment (mrwwheat7) [inline etcd-client/src/kv/mod.rs:71]: Another question is can we just use `cache_enable` here? When clean the cache, just set `cache_availability` to false, then after clean, set it to true

**Review by mrwwheat7 — CHANGES_REQUESTED** [review]

File: etcd-client/src/kv/mod.rs (hunk 5 above)
Comment (pzheng1025) [reply etcd-client/src/kv/mod.rs:71]: It is possible that taskA read `cache_enable=true` and execute something and switch to taskB, taskB finds cache error and set `cache_enable=false` and clean cache. And then switch back to taskA to continue, taskA will either read empty or read some intermediate status.

File: etcd-client/src/kv/mod.rs [hunk 6]
```diff
@@ -300,15 +371,23 @@ impl Kv {
     /// Will return `Err` if RPC call is failed.
     #[inline]
     pub async fn txn(&self, req: EtcdTxnRequest) -> Res<EtcdTxnResponse> {
-        if self.cache_enable && self.cache_err.load(Ordering::Acquire) {
-            return Err(EtcdError::InternalError("Kv cache meets an error".into()));
-        }
+        let inner = self.inner.read().await;
```
Comment (pzheng1025) [inline etcd-client/src/kv/mod.rs:374]: confirm read() is OK here.

**Review by pwang7 — APPROVED** [review]

**Review by mrwwheat7 — APPROVED** [review]

**Pull request merged**